### PR TITLE
Usr

### PR DESCRIPTION
**Pull request by bruntib**
This is a continuation of #420, so that pull request must be accepted first and this will be rebased on that.
TODO for this pull request:
- `mangledName` is still used at several places in the code, however, we have USR instead.
- There is a `generateUSRForType()` function in a later version of Clang. We need to upgrade Clang so we can use it instead of the current `getMangledName(QualType)` implementation.

### CODE REVIEW

**Comment by whisperity** [comment]
What if the following happens:

```cpp
// void use0() { f(-1); } -- cannot be done as symbol is not declared.

int f(int x);

void use1() { f(0); }
// CallExpr is linking the forward declaration

int f(int x) { return x; }

void use2() { f(1); }
// CallExpr is linking the definition
```

The two `CallExpr`s link two separate `FunctionDecl`s that are only reachable from one another because of a ReDecl chain.

Are we sure that the two functions will have USRs that are able to be found together? If not, we will break "jump to definition" across translation units, as calls from a file where in that TU only the "declaration" exists (due to headers) will only be able to jump to that symbol...

This should be tested.

File: plugins/cpp/parser/src/symbolhelper.cpp [hunk 1]
```diff
@@ -64,127 +54,20 @@ const clang::Type* getStrippedType(const clang::QualType& qt_)
...
-    // #else ... #endif section then every usage will also have two versions.
-    // So if a function has a template and a non-template version at the same
-    // time then querying the definition of the non-template usage the template
-    // version also returns.
-    return out.str();
-  }
-
-  return nd_->getQualifiedNameAsString();
+  llvm::SmallVector<char, 64> usr;
+  clang::index::generateUSRForDecl(nd_, usr);
```
Comment (whisperity) [inline plugins/cpp/parser/src/symbolhelper.cpp:60]: `generateUSRFor...` functions return a bool which are signalling whether "the result of the call should be ignored".

Comment (bruntib) [reply plugins/cpp/parser/src/symbolhelper.cpp:60]: Thanks, I haven't noticed the return type. I'll check what it means and in what cases it returns true or false.

**Comment by bruntib** [comment]
I strongly hope that all occurrences of function `f()` have the same USR in your code example. That's the goal of USR. It's a more interesting questions, what's the USR of the different `x` variables? This is the same question as in #427. Next week I'll have more time to investigate.

File: plugins/cpp/parser/src/symbolhelper.cpp (hunk 1 above)
Comment (bruntib) [reply plugins/cpp/parser/src/symbolhelper.cpp:84]: We investigated a little the return value of this function. I put a relatively long comment in the source code which describes why we don't use this value:
https://github.com/Ericsson/CodeCompass/pull/429/files#diff-b735f57c014d7a4293c17acf122b284dR60-R82
I don't know if this explanation is satisfactory.

**Comment by bruntib** [comment]
I changed "mangledName" to "entityHash" everywhere in the source code, because it is not a mangled name anymore.
I think I can remove WIP label from this pull request soon, because it approaches the final solution. Can anyone see any regression compared to the previous solution? Thank you!

**Comment by bruntib** [comment]
I'm brave enough now to remove WIP label from this patch. If you find some regression after using USR instead of our manually generated USR, feel free to comment it.

**Review by mcserep — APPROVED** [review]
While the CodeCompass self-parse test fails in the CI (on Ubuntu 18.04), it does not fail when tested on Ubuntu 16.04 and 18.04, on multiple computers.

I approve it, so it can fix #471 and we can get back to the CI issue later.

**Pull request merged**